### ISSUE TITLE
Bug: no-undef does not detect undefined 'this' properties

### ISSUE THREAD

**Issue by aromanet42** [eslint/eslint#15647]
### Environment

Node version: v17.3.0
npm version: v8.3.0
Local ESLint version: v8.9.0 (Currently used)
Global ESLint version: Not found
Operating System: linux 5.13.0-30-generic

### What parser are you using?

@babel/eslint-parser

### What did you do?

<details>
<summary>Configuration</summary>

[ESLint demo link](https://eslint.org/demo#eyJ0ZXh0IjoiY2xhc3MgU29tZXRoaW5nIHtcblxuICAgIHJlbmRlciAoKSB7XG5cbiAgICAgICAgcmV0dXJuIGAke3RoaXMuc3R1ZmZ9ICR7dW5kZWZpbmVkVmFyaWFibGV9YDtcblxuICAgIH1cblxufVxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJhY2Nlc3Nvci1wYWlycyI6MiwiYXJyYXktYnJhY2tldC1uZXdsaW5lIjoyLCJhcnJheS1icmFja2V0LXNwYWNpbmciOjIsImFycmF5LWNhbGxiYWNrLXJldHVybiI6MiwiYXJyYXktZWxlbWVudC1uZXdsaW5lIjoyLCJhcnJvdy1ib2R5LXN0eWxlIjoyLCJhcnJvdy1wYXJlbnMiOjIsImFycm93LXNwYWNpbmciOjIsImJsb2NrLXNjb3BlZC12YXIiOjIsImJsb2NrLXNwYWNpbmciOjIsImJyYWNlLXN0eWxlIjoyLCJjYWxsYmFjay1yZXR1cm4iOjIsImNhbWVsY2FzZSI6MiwiY2FwaXRhbGl6ZWQtY29tbWVudHMiOjIsImNsYXNzLW1ldGhvZHMtdXNlLXRoaXMiOjIsImNvbW1hLWRhbmdsZSI6MiwiY29tbWEtc3BhY2luZyI6MiwiY29tbWEtc3R5bGUiOjIsImNvbXBsZXhpdHkiOjIsImNvbXB1dGVkLXByb3BlcnR5LXNwYWNpbmciOjIsImNvbnNpc3RlbnQtcmV0dXJuIjoyLCJjb25zaXN0ZW50LXRoaXMiOjIsImNvbnN0cnVjdG9yLXN1cGVyIjoyLCJjdXJseSI6MiwiZGVmYXVsdC1jYXNlIjoyLCJkZWZhdWx0LWNhc2UtbGFzdCI6MiwiZGVmYXVsdC1wYXJhbS1sYXN0IjoyLCJkb3QtbG9jYXRpb24iOjIsImRvdC1ub3RhdGlvbiI6MiwiZW9sLWxhc3QiOjIsImVxZXFlcSI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZnVuYy1jYWxsLXNwYWNpbmciOjIsImZ1bmMtbmFtZS1tYXRjaGluZyI6MiwiZnVuYy1uYW1lcyI6MiwiZnVuYy1zdHlsZSI6MiwiZnVuY3Rpb24tY2FsbC1hcmd1bWVudC1uZXdsaW5lIjoyLCJmdW5jdGlvbi1wYXJlbi1uZXdsaW5lIjoyLCJnZW5lcmF0b3Itc3Rhci1zcGFjaW5nIjoyLCJnZXR0ZXItcmV0dXJuIjoyLCJnbG9iYWwtcmVxdWlyZSI6MiwiZ3JvdXBlZC1hY2Nlc3Nvci1wYWlycyI6MiwiZ3VhcmQtZm9yLWluIjoyLCJoYW5kbGUtY2FsbGJhY2stZXJyIjoyLCJpZC1ibGFja2xpc3QiOjIsImlkLWRlbnlsaXN0IjoyLCJpZC1sZW5ndGgiOjIsImlkLW1hdGNoIjoyLCJpbXBsaWNpdC1hcnJvdy1saW5lYnJlYWsiOjIsImluZGVudCI6MiwiaW5kZW50LWxlZ2FjeSI6MiwiaW5pdC1kZWNsYXJhdGlvbnMiOjIsImpzeC1xdW90ZXMiOjIsImtleS1zcGFjaW5nIjoyLCJrZXl3b3JkLXNwYWNpbmciOjIsImxpbmUtY29tbWVudC1wb3NpdGlvbiI6MiwibGluZWJyZWFrLXN0eWxlIjoyLCJsaW5lcy1hcm91bmQtY29tbWVudCI6MiwibGluZXMtYXJvdW5kLWRpcmVjdGl2ZSI6MiwibGluZXMtYmV0d2Vlbi1jbGFzcy1tZW1iZXJzIjoyLCJtYXgtY2xhc3Nlcy1wZXItZmlsZSI6MiwibWF4LWRlcHRoIjoyLCJtYXgtbGVuIjoyLCJtYXgtbGluZXMiOjIsIm1heC1saW5lcy1wZXItZnVuY3Rpb24iOjIsIm1heC1uZXN0ZWQtY2FsbGJhY2tzIjoyLCJtYXgtcGFyYW1zIjoyLCJtYXgtc3RhdGVtZW50cyI6MiwibWF4LXN0YXRlbWVudHMtcGVyLWxpbmUiOjIsIm11bHRpbGluZS1jb21tZW50LXN0eWxlIjoyLCJtdWx0aWxpbmUtdGVybmFyeSI6MiwibmV3LWNhcCI6MiwibmV3LXBhcmVucyI6MiwibmV3bGluZS1hZnRlci12YXIiOjIsIm5ld2xpbmUtYmVmb3JlLXJldHVybiI6MiwibmV3bGluZS1wZXItY2hhaW5lZC1jYWxsIjoyLCJuby1hbGVydCI6Miwibm8tYXJyYXktY29uc3RydWN0b3IiOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWF3YWl0LWluLWxvb3AiOjIsIm5vLWJpdHdpc2UiOjIsIm5vLWJ1ZmZlci1jb25zdHJ1Y3RvciI6Miwibm8tY2FsbGVyIjoyLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6Miwibm8tY2F0Y2gtc2hhZG93IjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25mdXNpbmctYXJyb3ciOjIsIm5vLWNvbnNvbGUiOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb25zdHJ1Y3Rvci1yZXR1cm4iOjIsIm5vLWNvbnRpbnVlIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZGl2LXJlZ2V4IjoyLCJuby1kdXBlLWFyZ3MiOjIsIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6Miwibm8tZHVwZS1lbHNlLWlmIjoyLCJuby1kdXBlLWtleXMiOjIsIm5vLWR1cGxpY2F0ZS1jYXNlIjoyLCJuby1kdXBsaWNhdGUtaW1wb3J0cyI6Miwibm8tZWxzZS1yZXR1cm4iOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LWZ1bmN0aW9uIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1lcS1udWxsIjoyLCJuby1ldmFsIjoyLCJuby1leC1hc3NpZ24iOjIsIm5vLWV4dGVuZC1uYXRpdmUiOjIsIm5vLWV4dHJhLWJpbmQiOjIsIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6Miwibm8tZXh0cmEtbGFiZWwiOjIsIm5vLWV4dHJhLXBhcmVucyI6Miwibm8tZXh0cmEtc2VtaSI6Miwibm8tZmFsbHRocm91Z2giOjIsIm5vLWZsb2F0aW5nLWRlY2ltYWwiOjIsIm5vLWZ1bmMtYXNzaWduIjoyLCJuby1nbG9iYWwtYXNzaWduIjoyLCJuby1pbXBsaWNpdC1jb2VyY2lvbiI6Miwibm8taW1wbGljaXQtZ2xvYmFscyI6Miwibm8taW1wbGllZC1ldmFsIjoyLCJuby1pbXBvcnQtYXNzaWduIjoyLCJuby1pbmxpbmUtY29tbWVudHMiOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWludmFsaWQtdGhpcyI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLWl0ZXJhdG9yIjoyLCJuby1sYWJlbC12YXIiOjIsIm5vLWxhYmVscyI6Miwibm8tbG9uZS1ibG9ja3MiOjIsIm5vLWxvbmVseS1pZiI6Miwibm8tbG9vcC1mdW5jIjoyLCJuby1sb3NzLW9mLXByZWNpc2lvbiI6Miwibm8tbWFnaWMtbnVtYmVycyI6Miwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLW1peGVkLW9wZXJhdG9ycyI6Miwibm8tbWl4ZWQtcmVxdWlyZXMiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbXVsdGktYXNzaWduIjoyLCJuby1tdWx0aS1zcGFjZXMiOjIsIm5vLW11bHRpLXN0ciI6Miwibm8tbXVsdGlwbGUtZW1wdHktbGluZXMiOjIsIm5vLW5hdGl2ZS1yZWFzc2lnbiI6Miwibm8tbmVnYXRlZC1jb25kaXRpb24iOjIsIm5vLW5lZ2F0ZWQtaW4tbGhzIjoyLCJuby1uZXN0ZWQtdGVybmFyeSI6Miwibm8tbmV3IjoyLCJuby1uZXctZnVuYyI6Miwibm8tbmV3LW9iamVjdCI6Miwibm8tbmV3LXJlcXVpcmUiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW5ldy13cmFwcGVycyI6Miwibm8tbm9ub2N0YWwtZGVjaW1hbC1lc2NhcGUiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLW9jdGFsLWVzY2FwZSI6Miwibm8tcGFyYW0tcmVhc3NpZ24iOjIsIm5vLXBhdGgtY29uY2F0IjoyLCJuby1wbHVzcGx1cyI6Miwibm8tcHJvY2Vzcy1lbnYiOjIsIm5vLXByb2Nlc3MtZXhpdCI6Miwibm8tcHJvbWlzZS1leGVjdXRvci1yZXR1cm4iOjIsIm5vLXByb3RvIjoyLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1yZXN0cmljdGVkLWV4cG9ydHMiOjIsIm5vLXJlc3RyaWN0ZWQtZ2xvYmFscyI6Miwibm8tcmVzdHJpY3RlZC1pbXBvcnRzIjoyLCJuby1yZXN0cmljdGVkLW1vZHVsZXMiOjIsIm5vLXJlc3RyaWN0ZWQtcHJvcGVydGllcyI6Miwibm8tcmVzdHJpY3RlZC1zeW50YXgiOjIsIm5vLXJldHVybi1hc3NpZ24iOjIsIm5vLXJldHVybi1hd2FpdCI6Miwibm8tc2NyaXB0LXVybCI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNlbGYtY29tcGFyZSI6Miwibm8tc2VxdWVuY2VzIjoyLCJuby1zZXR0ZXItcmV0dXJuIjoyLCJuby1zaGFkb3ciOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFjZWQtZnVuYyI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tc3luYyI6Miwibm8tdGFicyI6Miwibm8tdGVtcGxhdGUtY3VybHktaW4tc3RyaW5nIjoyLCJuby10ZXJuYXJ5IjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdGhyb3ctbGl0ZXJhbCI6Miwibm8tdHJhaWxpbmctc3BhY2VzIjoyLCJuby11bmRlZiI6Miwibm8tdW5kZWYtaW5pdCI6Miwibm8tdW5kZWZpbmVkIjoyLCJuby11bmRlcnNjb3JlLWRhbmdsZSI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVubW9kaWZpZWQtbG9vcC1jb25kaXRpb24iOjIsIm5vLXVubmVlZGVkLXRlcm5hcnkiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnJlYWNoYWJsZS1sb29wIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnNhZmUtb3B0aW9uYWwtY2hhaW5pbmciOjIsIm5vLXVudXNlZC1leHByZXNzaW9ucyI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXByaXZhdGUtY2xhc3MtbWVtYmVycyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZS1iZWZvcmUtZGVmaW5lIjoyLCJuby11c2VsZXNzLWJhY2tyZWZlcmVuY2UiOjIsIm5vLXVzZWxlc3MtY2FsbCI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1jb21wdXRlZC1rZXkiOjIsIm5vLXVzZWxlc3MtY29uY2F0IjoyLCJuby11c2VsZXNzLWNvbnN0cnVjdG9yIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8tdXNlbGVzcy1yZW5hbWUiOjIsIm5vLXVzZWxlc3MtcmV0dXJuIjoyLCJuby12YXIiOjIsIm5vLXZvaWQiOjIsIm5vLXdhcm5pbmctY29tbWVudHMiOjIsIm5vLXdoaXRlc3BhY2UtYmVmb3JlLXByb3BlcnR5IjoyLCJuby13aXRoIjoyLCJub25ibG9jay1zdGF0ZW1lbnQtYm9keS1wb3NpdGlvbiI6Miwib2JqZWN0LWN1cmx5LW5ld2xpbmUiOjIsIm9iamVjdC1jdXJseS1zcGFjaW5nIjoyLCJvYmplY3QtcHJvcGVydHktbmV3bGluZSI6Miwib2JqZWN0LXNob3J0aGFuZCI6Miwib25lLXZhciI6Miwib25lLXZhci1kZWNsYXJhdGlvbi1wZXItbGluZSI6Miwib3BlcmF0b3ItYXNzaWdubWVudCI6Miwib3BlcmF0b3ItbGluZWJyZWFrIjoyLCJwYWRkZWQtYmxvY2tzIjoyLCJwYWRkaW5nLWxpbmUtYmV0d2Vlbi1zdGF0ZW1lbnRzIjoyLCJwcmVmZXItYXJyb3ctY2FsbGJhY2siOjIsInByZWZlci1jb25zdCI6MiwicHJlZmVyLWRlc3RydWN0dXJpbmciOjIsInByZWZlci1leHBvbmVudGlhdGlvbi1vcGVyYXRvciI6MiwicHJlZmVyLW5hbWVkLWNhcHR1cmUtZ3JvdXAiOjIsInByZWZlci1udW1lcmljLWxpdGVyYWxzIjoyLCJwcmVmZXItb2JqZWN0LWhhcy1vd24iOjIsInByZWZlci1vYmplY3Qtc3ByZWFkIjoyLCJwcmVmZXItcHJvbWlzZS1yZWplY3QtZXJyb3JzIjoyLCJwcmVmZXItcmVmbGVjdCI6MiwicHJlZmVyLXJlZ2V4LWxpdGVyYWxzIjoyLCJwcmVmZXItcmVzdC1wYXJhbXMiOjIsInByZWZlci1zcHJlYWQiOjIsInByZWZlci10ZW1wbGF0ZSI6MiwicXVvdGUtcHJvcHMiOjIsInF1b3RlcyI6MiwicmFkaXgiOjIsInJlcXVpcmUtYXRvbWljLXVwZGF0ZXMiOjIsInJlcXVpcmUtYXdhaXQiOjIsInJlcXVpcmUtanNkb2MiOjIsInJlcXVpcmUtdW5pY29kZS1yZWdleHAiOjIsInJlcXVpcmUteWllbGQiOjIsInJlc3Qtc3ByZWFkLXNwYWNpbmciOjIsInNlbWkiOjIsInNlbWktc3BhY2luZyI6Miwic2VtaS1zdHlsZSI6Miwic29ydC1pbXBvcnRzIjoyLCJzb3J0LWtleXMiOjIsInNvcnQtdmFycyI6Miwic3BhY2UtYmVmb3JlLWJsb2NrcyI6Miwic3BhY2UtYmVmb3JlLWZ1bmN0aW9uLXBhcmVuIjoyLCJzcGFjZS1pbi1wYXJlbnMiOjIsInNwYWNlLWluZml4LW9wcyI6Miwic3BhY2UtdW5hcnktb3BzIjoyLCJzcGFjZWQtY29tbWVudCI6Miwic3RyaWN0IjoyLCJzd2l0Y2gtY29sb24tc3BhY2luZyI6Miwic3ltYm9sLWRlc2NyaXB0aW9uIjoyLCJ0ZW1wbGF0ZS1jdXJseS1zcGFjaW5nIjoyLCJ0ZW1wbGF0ZS10YWctc3BhY2luZyI6MiwidW5pY29kZS1ib20iOjIsInVzZS1pc25hbiI6MiwidmFsaWQtanNkb2MiOjIsInZhbGlkLXR5cGVvZiI6MiwidmFycy1vbi10b3AiOjIsIndyYXAtaWlmZSI6Miwid3JhcC1yZWdleCI6MiwieWllbGQtc3Rhci1zcGFjaW5nIjoyLCJ5b2RhIjoyfSwiZW52Ijp7fX19)

</details>

Minimal reproduction case:

```js
class Something {

    render () {

        return `${this.stuff} ${undefinedVariable}`;

    }

}
```

### What did you expect to happen?

I expected `this.stuff` to be reported as "not defined", just like `undefinedVariable`

### What actually happened?

`undefinedVariable` is indeed reported as "not defined"
`this.stuff` is NOT reported as "not defined"

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
Hi @aromanet42, thanks for the issue!

This works as intended because `no-undef` tracks only variables. We generally don't track properties as they can be defined on objects at any time.

For this purpose, you can use `tsc` ([demo](https://www.typescriptlang.org/play?filetype=js#code/MYGwhgzhAEDKD2BbApgFwBYEsB2BzaA3gFBHRnQBOy2AJshdABQCUhJ5HlaArhdtAAMAJAQyYIAOgipuAM1kBfaCO61ksnMhoA1MBUxgARiGQKBAbnbkFJBUA)).
